### PR TITLE
Remove backend cache from QiskitRuntimeService

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary

In #1715 we implemented new pattern to retrieve backend from the service, namely, opt-int. With opt-in, the same backend can have different Target, and thus they must be different Python objects.

This PR removes backend instance cache mechanism from `QiskitRuntimeService`. Instead, cache for the configuration data is implemented in the `RuntimeClient`. `IBMBackend` object is created every time it is requested, but the source data is cached and the communication with the server occurs only once.

### Details and comments

Once after we announce the fractional gate opt-in (not limited to fractional, some future opt-ins), user may try to compare the performance of our system with and without opt-in. One may write 

```python
backend_with_fg = service.backend("my_backend", use_fractional_gates=True)
backend_without_fg = service.backend("my_backend", use_fractional_gates=False)

# Run our system with fractional gates
sampler = SamplerV2(backend_with_fg)
isa_circ = transpile(qc, backend_with_fg)
result_with_fg = sampler.run(isa_circ).result()

# Run our system without fractional gates
sampler = SamplerV2(backend_without_fg)
isa_circ = transpile(qc, backend_without_fg)
result_without_fg = sampler.run(isa_circ).result()
```

Note that this does **NOT** work as expected BUT doesn't raise any warning or error. Note that `backend_with_fg` is silently mutated at the second retrieval via the cache, and both sampler jobs run without fractional gates. One may wonder why two results look comparable.

This PR will remove this complication and make internal code much simpler.

### CODE REVIEW

**Comment by jyu00** [comment]
I think there is value in caching the backend data, so we don't need to call the server again to retrieve the same exactly data just to build a target that with different basis gates. 

Having said that, the existing caching code is just too complicated, and this PR is a good opportunity to clean it up. A follow up improvement can be a "backend cache" that only stores server data (backend config etc), and`Backend` objects are dynamically created with different target, instance, etc but use the same server data.

File: qiskit_ibm_runtime/qiskit_runtime_service.py [hunk 1]
```diff
@@ -146,14 +142,11 @@ def __init__(
...
         self._channel = self._account.channel
-        self._backends: Dict[str, "ibm_backend.IBMBackend"] = {}
-        self._backend_configs: Dict[str, Any] = {}
+        self._backend_allowed_list: List[str] = []
 
         if self._channel == "ibm_cloud":
             self._api_client = RuntimeClient(self._client_params)
-            # TODO: We can make the backend discovery lazy
-            self._backends = self._discover_cloud_backends()
-            QiskitRuntimeService.global_service = self
```
[the remark this answers is not in the document]
Comment (nkanazawa1989) [reply qiskit_ibm_runtime/qiskit_runtime_service.py:156]: Thanks for catching this. I missed early return in this scope. Fixed in [6cb5f18](https://github.com/Qiskit/qiskit-ibm-runtime/pull/1732/commits/6cb5f181b8cbf420c70d655333ea157cb09687da)

File: qiskit_ibm_runtime/qiskit_runtime_service.py [hunk 2]
```diff
@@ -479,12 +456,13 @@ def _get_hgp(
...
+        for backend_name in self._discover_cloud_backends():
+            backend = self._create_backend_obj(backend_name)
+            attr_backend_name = to_python_identifier(backend_name)
             # Append _ if duplicate
-            while backend_name in self.__dict__:
-                backend_name += "_"
-            setattr(self, backend_name, backend)
+            while attr_backend_name in self.__dict__:
+                attr_backend_name += "_"
+            setattr(self, attr_backend_name, backend)
```
[the remark this answers is not in the document]
Comment (nkanazawa1989) [reply qiskit_ibm_runtime/qiskit_runtime_service.py]: Thanks for confirmation. I wondered why we have this never-used method.  Removed in [fea7de6](https://github.com/Qiskit/qiskit-ibm-runtime/pull/1732/commits/fea7de65c1f571a7288e3c7634843e7bf846877c)

**Comment by nkanazawa1989** [comment]
>  follow up improvement can be a "backend cache" that only stores server data (backend config etc), andBackend objects are dynamically created with different target, instance, etc but use the same server data.

Done in [deb6cb6](https://github.com/Qiskit/qiskit-ibm-runtime/pull/1732/commits/deb6cb653e812651a46409bcdb1a2d99cc419c1f). Cache code is now very simple.

**Review by jyu00 — APPROVED** [review]
LGTM!

**Comment by jyu00** [comment]
Actually, one more comment - Both `backend.properties()` and `backend.defaults()` currently have a `refresh` keyword that says "If ``True``, re-query the server for the backend properties/defaults. Otherwise, return a cached version.". Since we only cache configuration now, those need to be deprecated and later removed, unless you want to also add caching for properties/defaults.

**Comment by nkanazawa1989** [comment]
> Since we only cache configuration now, those need to be deprecated and later removed, unless you want to also add caching for properties/defaults.

I think this behavior should be okey because instance-wise these information are cached. Maybe I should write upgrade note.

**Pull request merged**